src/System.Private.Uri/src/System/UriHelper.cs

```diff
@@ -265,17 +299,17 @@ internal static class UriHelper
         // - It is a RARE case when Unescape actually needs escaping some characters mentioned above.
         //   For this reason it returns a char[] that is usually the same ref as the input "dest" value.
         //
-        internal static unsafe char[] UnescapeString(string input, int start, int end, char[] dest,
+        internal static unsafe ValueStringBuilder UnescapeString(string input, int start, int end, ref ValueStringBuilder dest,
             ref int destPosition, char rsvd1, char rsvd2, char rsvd3, UnescapeMode unescapeMode, UriParser? syntax,
             bool isQuery)
         {
             fixed (char* pStr = input)
             {
-                return UnescapeString(pStr, start, end, dest, ref destPosition, rsvd1, rsvd2, rsvd3, unescapeMode,
+                return UnescapeString(pStr, start, end, ref dest, ref destPosition, rsvd1, rsvd2, rsvd3, unescapeMode,
                     syntax, isQuery);
             }
         }
-        internal static unsafe char[] UnescapeString(char* pStr, int start, int end, char[] dest, ref int destPosition,
+        internal static unsafe ValueStringBuilder UnescapeString(char* pStr, int start, int end, ref ValueStringBuilder dest, ref int destPosition,
```

Comment:
We should not be returning ValueStringBuilder.  That makes a copy and is dangerous.

Reply:
Fixed